cmake/FindSTM32HAL.cmake

```diff
@@ -1,176 +1,190 @@
     ENDIF()
-ENDFOREACH()
-
-LIST(REMOVE_DUPLICATES HAL_HEADERS)
-LIST(REMOVE_DUPLICATES HAL_SRCS)
-
-STRING(TOLOWER ${STM32_FAMILY} STM32_FAMILY_LOWER)
-
-FIND_PATH(STM32HAL_INCLUDE_DIR ${HAL_HEADERS}
-    PATH_SUFFIXES include stm32${STM32_FAMILY_LOWER}
-    HINTS ${STM32Cube_DIR}/Drivers/STM32${STM32_FAMILY}xx_HAL_Driver/Inc
-    CMAKE_FIND_ROOT_PATH_BOTH
-)
-
-FOREACH(HAL_SRC ${HAL_SRCS})
-    STRING(MAKE_C_IDENTIFIER "${HAL_SRC}" HAL_SRC_CLEAN)
-    SET(HAL_${HAL_SRC_CLEAN}_FILE HAL_SRC_FILE-NOTFOUND)
-    FIND_FILE(HAL_${HAL_SRC_CLEAN}_FILE ${HAL_SRC}
-        PATH_SUFFIXES src stm32${STM32_FAMILY_LOWER}
-        HINTS ${STM32Cube_DIR}/Drivers/STM32${STM32_FAMILY}xx_HAL_Driver/Src
+    
+    FOREACH(cmp ${HAL_REQUIRED_COMPONENTS})
+        LIST(FIND STM32HAL_FIND_COMPONENTS ${cmp} STM32HAL_FOUND_INDEX)
+        IF(${STM32HAL_FOUND_INDEX} LESS 0)
+            LIST(APPEND STM32HAL_FIND_COMPONENTS ${cmp})
+        ENDIF()
+    ENDFOREACH()
+    
+    FOREACH(cmp ${STM32HAL_FIND_COMPONENTS})
+        LIST(FIND HAL_COMPONENTS ${cmp} STM32HAL_FOUND_INDEX)
+        IF(${STM32HAL_FOUND_INDEX} LESS 0)
+            MESSAGE(FATAL_ERROR "Unknown STM32HAL component: ${cmp}. Available components: ${HAL_COMPONENTS}")
+    	ELSE()
+            LIST(APPEND HAL_HEADERS ${HAL_PREFIX}hal_${cmp}.h)
+            LIST(APPEND HAL_SRCS ${HAL_PREFIX}hal_${cmp}.c)
+        ENDIF()
+        LIST(FIND HAL_EX_COMPONENTS ${cmp} STM32HAL_FOUND_INDEX)
+        IF(NOT (${STM32HAL_FOUND_INDEX} LESS 0))
+            LIST(APPEND HAL_HEADERS ${HAL_PREFIX}hal_${cmp}_ex.h)
+            LIST(APPEND HAL_SRCS ${HAL_PREFIX}hal_${cmp}_ex.c)
+        ENDIF()
+    ENDFOREACH()
+    
+    LIST(REMOVE_DUPLICATES HAL_HEADERS)
+    LIST(REMOVE_DUPLICATES HAL_SRCS)
+    
+    STRING(TOLOWER ${STM32_FAMILY} STM32_FAMILY_LOWER)
+    
+    FIND_PATH(STM32HAL_INCLUDE_DIR ${HAL_HEADERS}
+        PATH_SUFFIXES include stm32${STM32_FAMILY_LOWER}
+        HINTS ${STM32Cube_DIR}/Drivers/STM32${STM32_FAMILY}xx_HAL_Driver/Inc
         CMAKE_FIND_ROOT_PATH_BOTH
     )
-    LIST(APPEND STM32HAL_SOURCES ${HAL_${HAL_SRC_CLEAN}_FILE})
-ENDFOREACH()
+    
+    FOREACH(HAL_SRC ${HAL_SRCS})
+        STRING(MAKE_C_IDENTIFIER "${HAL_SRC}" HAL_SRC_CLEAN)
+        SET(HAL_${HAL_SRC_CLEAN}_FILE HAL_SRC_FILE-NOTFOUND)
+        FIND_FILE(HAL_${HAL_SRC_CLEAN}_FILE ${HAL_SRC}
+            PATH_SUFFIXES src stm32${STM32_FAMILY_LOWER}
+            HINTS ${STM32Cube_DIR}/Drivers/STM32${STM32_FAMILY}xx_HAL_Driver/Src
+            CMAKE_FIND_ROOT_PATH_BOTH
+        )
+        LIST(APPEND STM32HAL_SOURCES ${HAL_${HAL_SRC_CLEAN}_FILE})
+    ENDFOREACH()
+ENDIF()
 
 INCLUDE(FindPackageHandleStandardArgs)
 
 FIND_PACKAGE_HANDLE_STANDARD_ARGS(STM32HAL DEFAULT_MSG STM32HAL_INCLUDE_DIR STM32HAL_SOURCES)
+
+if(STM32HAL_FOUND AND NOT TARGET STM32Cube::HAL)
+    add_library(STM32Cube::HAL INTERFACE IMPORTED)
+    set_target_properties(STM32Cube::HAL 
```

Comment:
better use the target_blabla() functions to set target properties

Reply:
See https://github.com/ObKo/stm32-cmake/pull/87#discussion_r322024699